src/UI/Components/Diagram/helpers.ts

```diff
@@ -220,3 +221,109 @@ const doesElementsIsEmbeddedWithExhaustedConnections = (
   }
   return false;
 };
+
+/**
+ * Function that will merge state from Instance Composer to proper object for order_api endpoint.
+ * Instance composer state is being split into multiple objects that could be embedded into other available, so we need to recursively
+ * go through all of them to group, and sort them
+ * @param instances all of the instances that were created/edited in the instance, not including the one passed in second parameter
+ * @param instance instance which is used taken into consideration as the parent of the possible embedded
+ * @param isEmbedded boolean informing whether instance passed is embedded or not
+ * @returns
+ */
+export const shapesDataTransform = (
+  instances: InstanceForApi[],
+  instance: InstanceForApi,
+  isEmbedded = false,
+) => {
+  let areEmbeddedEdited = false;
+  const matchingInstances = instances.filter(
+    (checkedInstance) => checkedInstance.embeddedTo === instance.instance_id,
+  );
+
+  const notMatchingInstances = instances.filter(
+    (checkedInstance) => checkedInstance.embeddedTo !== instance.instance_id,
+  );
+
+  //iterate through matching (embedded)instances and group them according to property type to be able to put them in the Array if needed at once
+  const groupedEmbedded: { [key: string]: InstanceForApi[] } =
+    matchingInstances.reduce(function (r, a) {
+      r[a.service_entity] = r[a.service_entity] || [];
+      r[a.service_entity].push(a);
+      return r;
+    }, Object.create({}));
+
+  for (const [key, instancesToEmbed] of Object.entries(groupedEmbedded)) {
+    if (instance.value) {
+      if (instancesToEmbed.length > 1) {
+        const updated: { [key: string]: unknown }[] = [];
+        instancesToEmbed.forEach((instance) => {
+          const updatedInstance = shapesDataTransform(
+            notMatchingInstances,
+            instance,
+            true,
+          );
+
+          areEmbeddedEdited =
+            areEmbeddedEdited !== true &&
+            instance.action === null &&
+            (updatedInstance.action !== null ||
```

Comment:
you got three actions, create, update, delete, so if it's not null, it's one of the three. So your first check cancels out the second. to be of type "delete" it's by definition, not null. Since you used an or statement, it will never bother doing the "delete" check if the first statement is valid. In the case the first statement is invalid, it means that the action is null, and thus can never be "delete".

Reply:
I had to get lost in those conditional statements, thank for pointing it out!